docs/proposed/http-storage-node-protocol.rst

```diff
@@ -0,0 +1,506 @@
+Imagine the SPKI hash is ``i5xb...``.
+This results in a fURL of ``pb://i5xb...@example.com:443/g3m5...#v=2`` [#]_.
+Bob creates a client node pointed at the same introducer.
+Bob's client node receives the announcement from Alice's storage node.
+
+Bob's client node recognizes the fURL as referring to an HTTP-dialect server due to the ``v=2`` fragment.
+Bob's client node can now perform a TLS handshake with a server at the address in the fURL location hints
+(``example.com:443`` in this example).
+Following the above described validation procedures,
+Bob's client node can determine whether it has reached Alice's storage node or not.
+If and only if the validation procedure is successful does Bob's client node conclude it has reached Alice's storage node.
+**Peer authentication** has been achieved.
+
+Additionally,
+by continuing to interact using TLS,
+Bob's client and Alice's storage node are assured of both **message authentication** and **message confidentiality**.
+
+.. note::
+
+   Foolscap TubIDs are 20 bytes (SHA1 digest of the certificate).
+   They are presented with base32 encoding at a length of 32 bytes.
+   SPKI information discussed here is 32 bytes (SHA256 digest).
+   They will present in base32 as 52 bytes.
+   https://tools.ietf.org/html/rfc7515#appendix-C may prove a better (more compact) choice for encoding the information into a fURL.
+   It will encode 32 bytes into merely 43...
+   We could also choose to reduce the hash size of the SPKI information through use of another cryptographic hash (replacing sha256).
+   A 224 bit hash function (SHA3-224, for example) might be suitable -
+   improving the encoded length to 38 bytes.
+   Or we could stick with the Foolscap digest function - SHA1.
+
+
+Transition
+~~~~~~~~~~
+
+To provide a seamless user experience during this protocol transition,
+there should be a period during which both protocols are supported by storage nodes.
+The GBS announcement will be introduced in a way that *updated client* software can recognize.
+Its introduction will also be made in such a way that *non-updated client* software disregards the new information
+(of which it cannot make any use).
+
+Storage nodes will begin to operate a new GBS server.
+They may re-use their existing x509 certificate or generate a new one.
+Generation of a new certificate allows for certain non-optimal conditions to be addressed:
+
+* The ``commonName`` of ``newpb_thingy`` may be changed to a more descriptive value.
+* A ``notValidAfter`` field with a timestamp in the past may be updated.
+
+Storage nodes will announce a new fURL for this new HTTP-based server.
+This fURL will be announced alongside their existing Foolscap-based server's fURL.
+Such an announcement will resemble this::
+
+  {
+      "anonymous-storage-FURL": "pb://...",          # The old key
+      "anonymous-storage-gbs-FURL": "pb://...#v=2"   # The new key
+  }
```

Comment:
I feel like we shouldn't have upper-case stuff in these keys, now's our chance to kill it? ;)
(Will we ever have more GPS things to announce, maybe ``gbs-anonymous-storage-furl`` so other GPS things can be ``gbs-`` .. but right into bikeshed territory)